File: examples/dekita/.claude/hooks/lib/path_validation.py
#!/usr/bin/env python3
"""パストラバーサル防止のためのパス検証ユーティリティを提供する。

Why:
    外部入力（transcript path等）のパストラバーサル攻撃を防止し、
    許可されたディレクトリのみへのアクセスを保証する。

What:
    - is_safe_transcript_path(): transcriptパスの安全性検証

Remarks:
    - 許可ディレクトリ: ホーム、システムtemp、cwd
    - シンボリックリンクは解決後に検証
    - Claude Code内部生成パスでも防御的に検証

Changelog:
    - silenvx/dekita#1914: セキュリティレビュー対応で追加
"""

import os
import tempfile
from pathlib import Path


def is_safe_transcript_path(path_str: str) -> bool:
    """Validate that a transcript path is safe to read.

    Checks that the path:
    1. Is not empty
    2. Is absolute or can be safely resolved
    3. After resolution, is within allowed directories:
       - User's home directory
       - System temp directories
       - Current working directory

    Args:
        path_str: The path string to validate.

    Returns:
        True if the path is safe, False otherwise.

    Note:
        This function is designed for validating transcript paths provided
        by Claude Code's hook system. In practice, these paths are generated
        by Claude Code itself, but validation is good defensive practice.
    """
    if not path_str or not path_str.strip():
        return False

    try:
        # Resolve to absolute path, following symlinks
        resolved = Path(path_str).resolve()

        # Get allowed directories
        allowed_dirs = _get_allowed_directories()

        # Check if resolved path is under any allowed directory
        return any(_is_path_under(resolved, allowed_dir) for allowed_dir in allowed_dirs)
    except (OSError, ValueError):
        # Path resolution failed (e.g., invalid characters)
        return False


def _get_allowed_directories() -> list[Path]:
    """Get list of allowed directories for transcript files.

    Returns:
        List of Path objects representing allowed directories.
    """
    allowed = []

    # User's home directory
    home = Path.home()
    if home.exists():
        allowed.append(home)

    # System temp directories
    tmpdir = os.environ.get("TMPDIR")
    if tmpdir:
        tmp_path = Path(tmpdir)
        if tmp_path.exists():
            allowed.append(tmp_path)

    # System temp directory (cross-platform)
    system_tmp = Path(tempfile.gettempdir())
    if system_tmp.exists() and system_tmp not in allowed:
        allowed.append(system_tmp)

    # Current working directory (for relative paths)
    try:
        cwd = Path.cwd()
        if cwd.exists():
            allowed.append(cwd)
    except OSError:
        pass  # CWD may not exist in edge cases

    return allowed


def _is_path_under(path: Path, directory: Path) -> bool:
    """Check if path is under directory (handles symlinks).

    Args:
        path: The resolved path to check.
        directory: The directory to check against.

    Returns:
        True if path is under directory, False otherwise.
    """
    try:
        resolved_dir = directory.resolve()
        # Use is_relative_to for clean comparison (Python 3.9+)
        return path.is_relative_to(resolved_dir)
    except (OSError, ValueError):
        return False

File: examples/dekita/.claude/hooks/lib/path_validation.ts
/**
 * パストラバーサル防止のためのパス検証ユーティリティを提供する。
 *
 * Why:
 *   外部入力（transcript path等）のパストラバーサル攻撃を防止し、
 *   許可されたディレクトリのみへのアクセスを保証する。
 *
 * What:
 *   - isSafeTranscriptPath(): transcriptパスの安全性検証
 *
 * Remarks:
 *   - 許可ディレクトリ: ホーム、システムtemp、cwd
 *   - シンボリックリンクは解決後に検証
 *   - Claude Code内部生成パスでも防御的に検証
 *
 * Changelog:
 *   - silenvx/dekita#2868: Python版から移行
 */

import { existsSync, realpathSync } from "node:fs";
import { homedir, tmpdir } from "node:os";
import { relative, resolve } from "node:path";

/**
 * Get list of allowed directories for transcript files.
 *
 * @returns Array of resolved directory paths.
 */
function getAllowedDirectories(): string[] {
  const allowed: string[] = [];

  // User's home directory
  try {
    const home = homedir();
    if (existsSync(home)) {
      allowed.push(realpathSync(home));
    }
  } catch {
    // Ignore errors
  }

  // System temp directories
  const tmpdirEnv = process.env.TMPDIR;
  if (tmpdirEnv) {
    try {
      if (existsSync(tmpdirEnv)) {
        const resolved = realpathSync(tmpdirEnv);
        if (!allowed.includes(resolved)) {
          allowed.push(resolved);
        }
      }
    } catch {
      // Ignore errors
    }
  }

  // System temp directory (cross-platform)
  try {
    const systemTmp = tmpdir();
    if (existsSync(systemTmp)) {
      const resolved = realpathSync(systemTmp);
      if (!allowed.includes(resolved)) {
        allowed.push(resolved);
      }
    }
  } catch {
    // Ignore errors
  }

  // Current working directory (for relative paths)
  try {
    const cwd = process.cwd();
    if (existsSync(cwd)) {
      const resolved = realpathSync(cwd);
      if (!allowed.includes(resolved)) {
        allowed.push(resolved);
      }
    }
  } catch {
    // CWD may not exist in edge cases
  }

  return allowed;
}

/**
 * Check if path is under directory (handles symlinks).
 *
 * Uses relative path calculation to safely determine containment,
 * avoiding startsWith() which is vulnerable to path traversal.
 *
 * @param path - The resolved path to check.
 * @param directory - The directory to check against.
 * @returns True if path is under directory, false otherwise.
 */
function isPathUnder(pathStr: string, directory: string): boolean {
  try {
    const resolvedDir = realpathSync(resolve(directory));
    const resolvedPath = realpathSync(resolve(pathStr));

    // Use relative path to check containment
    // If path is under directory, relative will not start with '..' or '/'
    const rel = relative(resolvedDir, resolvedPath);

    // Path is under directory if:
    // - relative path doesn't start with '..' (not escaping directory)
    // - relative path doesn't start with '/' (not absolute - shouldn't happen after resolve)
    // - relative path is not empty (path == directory is allowed)
    return !rel.startsWith("..") && !rel.startsWith("/");
  } catch {
    return false;
  }
}

/**
 * Validate that a transcript path is safe to read.
 *
 * Checks that the path:
 * 1. Is not empty
 * 2. Is absolute or can be safely resolved
 * 3. After resolution, is within allowed directories:
 *    - User's home directory
 *    - System temp directories
 *    - Current working directory
 *
 * @param pathStr - The path string to validate.
 * @returns True if the path is safe, false otherwise.
 *
 * Note:
 *   This function is designed for validating transcript paths provided
 *   by Claude Code's hook system. In practice, these paths are generated
 *   by Claude Code itself, but validation is good defensive practice.
 */
export function isSafeTranscriptPath(pathStr: string): boolean {
  if (!pathStr || !pathStr.trim()) {
    return false;
  }

  try {
    // Resolve to absolute path, following symlinks
    const resolved = realpathSync(resolve(pathStr));

    // Get allowed directories
    const allowedDirs = getAllowedDirectories();

    // Check if resolved path is under any allowed directory
    return allowedDirs.some((allowedDir) => isPathUnder(resolved, allowedDir));
  } catch {
    // Path resolution failed (e.g., file doesn't exist, invalid characters)
    return false;
  }
}
